### DISCUSSION TITLE
ForFhir<>() requires usage of deprecated constructors

### DISCUSSION THREAD

**Discussion by FullStackChef** [FirelyTeam/firely-net-sdk#2954] [Q&A]
Hi, we previously had our own implementation of the FhirJsonConverterFactory and FhirJsonConverter which extended the standard one in order to performa few tasks that make sense to do during the serialization process

In the most recent versions all of the constructors have been deprecated with a message recommending the use of ForFhir() 
We have switched to doing so with ForFhir() for our general serialization and am trying to use ForFhir<Type>() for our custom converters
however ForFhir<Type>() expects an argument of type FhirJsonConverter<T> which in order to use we have to use a constructor that as above is marked as deprecated.

I also can't find any overload of ForFhir that takes a serializationfilter as per the documentation. I'm struggling to understand what the correct approach is and note that the documentation is fairly light in this area - I understand that this is probably a bit of an edge case.

What is the correct / recommended approach is feature complete, I'm happy to contribute to issues / documentation if required.

**Comment by Kasdejong** [comment]
Hi! Could you maybe provide a little information about what extra steps you are taking? We deprecated these methods because we wanted users to use `JsonSerializerOptions.ForFhir()` instead, not `FhirJsonConverter.ForFhir()`. Maybe we should have documented this better!

In any case, we should discuss internally because you are most likely not the only one to do this. Right now, our intended solution for users that want to override the default behaviour is for them to subclass `BaseFhirJsonPocoSerializer` and `BaseFhirJsonPocoDeserializer`, then use those directly.

We are aware that our documentation is running behind :)

**Reply by FullStackChef** [reply]
Hopefully these images might help :)
![{12693BA7-FD4B-406D-B4B5-D07523F193B9}](https://github.com/user-attachments/assets/8cb44641-b668-4400-aa2e-c38cd87b8d2b)
Note in the image above I am using JsonSerializerOptions.ForFhir() (just with a type) which according to the source looks like a valid approach - however the ForFhir<T> doesn't take a type of JsonConverter<T> it takes a type of FhirJsonConverter<T> which means that my custom converter needs to inherit from the FhirJsonConverter

The code below is contained in an abstract converter that we use - (for one it means I can ignore the deprecated warnings in one place) but also we can centralize some logic per the image that we want to run everytime we serialize. Attachment processing is one of these - our consuming applications require html rather than rtf so we make the conversion here - Ideally we would just write a custom serializer for attachment but due to the inner workings of the fhir serializer that would never get called and ultimately this performs well enough.
With the FhirURI we use the SDK in a .net application that acts as a facade for our fhir server in Azure - it hadles things like security etc.  Any URI coming out of the Azure Health services has a base url of the actual fhir server - this application sits behind an application gateway and we need to ensure that those urls get transformed to point to the public endpoint

![{1C353A92-33C2-43DB-8EFE-1FAF58A80635}](https://github.com/user-attachments/assets/0e828924-6e96-4498-9939-5bda58e6fb7d)

Perhaps we need to adjust our approach to work more directly with the BaseFhirJsonPocoDeserializer however my assumptions was that we should be tapping into the converter level. 

As mentioned, - happy to contribute to documentation if needed.

Cheers,
